packages/indexer/src/core/rateLimiting.ts

```diff
@@ -0,0 +1,177 @@
+  cupsPerGetTransactionReceipt: number;
+  cupsPerGetBlockNumber: number;
+  cupsPerGetLogs: number;
+}
+
+export interface RPPLimiterStrategy {
+  requests: number;
+  ms: number;
+}
+
+export const DefaultCups = {
+  cupsPerGetBlock: 16,
+  cupsPerGetTransaction: 17,
+  cupsPerGetTransactionReceipt: 15,
+  cupsPerGetBlockNumber: 10,
+  cupsPerGetLogs: 75,
+};
+
+export const infuraDefaultLimiterStrategyDaily: RPPLimiterStrategy = {
+  requests: 100_000,
+  ms: 24 * 60 * 60 * 1000,
+};
+
+export const xxDefaultLimiterStrategy: CupsLimiterStrategy = {
+  cu: 1.15 * SINGLE_REQUEST_CU,
+  ms: 1000,
+  ...DefaultCups,
+};
+
+export const infuraDefaultLimiterStrategy: CupsLimiterStrategy = {
+  cu: 1.15 * SINGLE_REQUEST_CU,
+  ms: 1000,
+  ...DefaultCups,
+};
+
+export const alchemyDefaultLimiterStrategy: CupsLimiterStrategy = {
+  cu: 330,
+  ms: 1000,
+  cupsPerGetBlock: 16,
+  cupsPerGetTransaction: 17,
+  cupsPerGetTransactionReceipt: 15,
+  cupsPerGetBlockNumber: 10,
+  cupsPerGetLogs: 75,
+};
+
+const MAX_CUPS = 75;
+
+export class RPCRateLimiter {
+  limiter: RateLimiter;
+  strategy: CupsLimiterStrategy;
+  constructor(strategy: CupsLimiterStrategy | RPPLimiterStrategy) {
+    if ('cu' in strategy) {
+      this.strategy = strategy;
+    } else if ('requests' in strategy) {
+      this.strategy = {
+        cu: strategy.requests,
+        ms: strategy.ms,
+        cupsPerGetBlock: 16,
+        cupsPerGetTransaction: 17,
+        cupsPerGetTransactionReceipt: 15,
+        cupsPerGetBlockNumber: 10,
+        cupsPerGetLogs: 75,
+      };
+    } else {
+      throw new Error(`Limiting strategy undefined`);
+    }
+
+    const multiplier = Math.ceil(MAX_CUPS / this.strategy.cu);
+    const tokensPerInterval = MAX_CUPS * multiplier;
+    const interval = this.strategy.ms * multiplier;
+
+    this.limiter = new RateLimiter({ tokensPerInterval, interval });
+  }
+
+  async getBlockWithTransactions() {
+    await this.limiter.removeTokens(this.strategy.cupsPerGetBlock || 1);
+  }
+
+  async getBlock() {
+    await this.limiter.removeTokens(this.strategy.cupsPerGetBlock);
+  }
+
+  async getTransaction() {
+    await this.limiter.removeTokens(this.strategy.cupsPerGetTransaction);
+  }
+
+  async getTransactionReceipt() {
+    await this.limiter.removeTokens(this.strategy.cupsPerGetTransactionReceipt);
+  }
+
+  async getBlockNumber() {
+    await this.limiter.removeTokens(this.strategy.cupsPerGetBlockNumber);
+  }
+
+  async getLogs() {
+    await this.limiter.removeTokens(this.strategy.cupsPerGetLogs);
+  }
+}
+
+export function strategyFromRPP(rps: number, secs = 1): CupsLimiterStrategy {
+  return {
+    cu: rps * SINGLE_REQUEST_CU,
+    ms: secs * 1000,
+    ...DefaultCups,
+  };
+}
+
+export function tryLimitStrategyByName(
+  name: string,
+): CupsLimiterStrategy | RPPLimiterStrategy | undefined {
+  const _name = name.toLowerCase();
+  switch (_name) {
+    case 'moonbeam':
+      return strategyFromRPP(3.9); // 1.9
```

Comment:
Where do these numbers come from?

Reply:
good question! They are coming from the calculation of the relative weighted PRC requests per second. For example, [alchemy allows us to do 330 compute units (CU) per second](https://docs.alchemy.com/alchemy/documentation/throughput), where every method call has a fixed CU. ﻿﻿eth_blockNumber costs 10 CUs, eth_getLogs is 75. If we make every second 10 requests by 10 CUs and 10 by 75, then in total we make 850 CUs per second. In reality this number of allowed CU's/sec is waaay lower. (Also the approximate CU's are calculated for Alchemy, but we use the same numbers assuming that other RPC have somewhat similar mechanisms)

So having this knowledge I calculated the approx CU/sec for every RPC we run using [Graphana data](https://nomadxyz.grafana.net/d/hxT-q6-7z/indexer-production?orgId=1&var-network=ethereum&var-environment=production&viewPanel=42) and [spread sheets](https://docs.google.com/spreadsheets/d/16vicOJwCqMHHWZ1pmvw8DQYEGXTNs1uXQFrMszZ8694/edit#gid=0). The numbers I've got are in the // comment, and I fine-tuned them a bit, so u see 3.9 as the tuned number, which works alright, because alchemy works a bit differently to others. We should make this fine-tuning automatically using RPC tester which is in @yourbuddyconner 's plans for long time.